tests/test_p2p_daemon.py

```diff
@@ -0,0 +1,296 @@
+import asyncio
+import multiprocessing as mp
+import subprocess
+from functools import partial
+
+from hivemind.p2p.p2p_daemon_bindings.datastructures import ID
+
+import numpy as np
+import pytest
+
+from hivemind.p2p import P2P
+from hivemind.proto import dht_pb2
+
+RUNNING = 'running'
+NOT_RUNNING = 'not running'
+CHECK_PID_CMD = '''
+if ps -p {0} > /dev/null;
+then
+    echo "{1}"
+else
+    echo "{2}"
+fi
+'''
+
+
+def is_process_running(pid: int) -> bool:
+    cmd = CHECK_PID_CMD.format(pid, RUNNING, NOT_RUNNING)
+    return subprocess.check_output(cmd, shell=True).decode('utf-8').strip() == RUNNING
+
+
+async def replicate_if_needed(p2p: P2P, replicate: bool):
+    return await P2P.replicate(p2p._daemon_listen_port, p2p._host_port) if replicate else p2p
+
+
+@pytest.mark.asyncio
+async def test_daemon_killed_on_del():
+    p2p_daemon = await P2P.create()
+
+    child_pid = p2p_daemon._child.pid
+    assert is_process_running(child_pid)
+
+    p2p_daemon.__del__()
```

Comment:
[note]
make sure we do not call __del__ directly to avoid double calls 
(how __del__ works: https://stackoverflow.com/a/1481512 )